backend/wallet/services/transaction.py

```diff
@@ -124,16 +150,95 @@ def process_incoming_transaction(
             f" but received transaction has amount: {amount}, status: {transaction.status}, "
             f"source address: {sender_address}, destination address: {receiver_address}"
         )
 
-    if not receiver_id:
-        log_execution("Incoming transaction had no metadata. crediting inventory")
-        receiver_id = get_account(account_name=INVENTORY_ACCOUNT_NAME).id
+    if isinstance(metadata, diem_types.Metadata__RefundMetadata) and isinstance(
+        metadata.value, diem_types.RefundMetadata__RefundMetadataV0
+    ):
+        logger.info(
+            f"process_incoming_transaction refund metadata [{blockchain_version}]: "
+            f"sender: {sender_address}, receiver: {receiver_address}, "
+            f"seq: {sequence}, amount: {amount}"
+        )
+        txn_version = metadata.value.value.transaction_version
+        reason = metadata.value.value.reason
+        payment_type = TransactionType.REFUND
+
+        original_txn = get_transaction_by_blockchain_version(txn_version)
+
+        logger.info(f"Refund metadata: txn version: {txn_version}, reason: {reason}")
+
+        # Cannot find the referred original transaction
+        if original_txn is None:
+            logger.info(
+                f"Refund metadata invalid: txn version: {txn_version} does not exist"
+                f"Sending refund back to sender {sender_address}"
+            )
+            # Record the received transaction as a regular transaction to inventory account
+            inventory_account_id = get_account(account_name=INVENTORY_ACCOUNT_NAME).id
+            add_transaction(
+                amount=amount,
+                currency=currency,
+                payment_type=TransactionType.EXTERNAL,
+                status=TransactionStatus.FAILED,
+                source_address=sender_address,
+                destination_id=inventory_account_id,
+                destination_address=receiver_address,
+                sequence=sequence,
+                blockchain_version=blockchain_version,
+            )
+            return
+
+        if isinstance(reason, diem_types.RefundReason__InvalidSubaddress):
+            refund_reason = RefundReason.INVALID_SUBADDRESS
+        elif isinstance(reason, diem_types.RefundReason__UserInitiatedPartialRefund):
+            refund_reason = RefundReason.USER_INITIATED_PARTIAL_REFUND
+        elif isinstance(reason, diem_types.RefundReason__UserInitiatedFullRefund):
+            refund_reason = RefundReason.USER_INITIATED_FULL_REFUND
+        else:
+            refund_reason = RefundReason.OTHER
+
+        original_txn = get_transaction_by_blockchain_version(txn_version)
+        receiver_id = original_txn.source_id
+        receiver_subaddr = original_txn.source_subaddress
+        sender_subaddress = original_txn.destination_subaddress
+        original_txn_id = original_txn.id
+
+    # Could not find receiver for the given subaddress, so send back a refund from inventory account
+    if receiver_id is None:
```

Comment:
it is really hard to reason why here you get receiver_id is None, because above code:

```
        if general_v0.to_subaddress:
            receiver_subaddr = general_v0.to_subaddress.hex()
            receiver_id = get_account_id_from_subaddr(receiver_subaddr)
        else:
            # Subaddress was not specified, so route the fund to inventory account
            receiver_id = get_account(account_name=INVENTORY_ACCOUNT_NAME).id
```
It looks like the reason is because get_account_id_from_subaddr returns None if not found, for that case, could we do:

```
        if general_v0.to_subaddress:
            receiver_subaddr = general_v0.to_subaddress.hex()
            receiver_id = get_account_id_from_subaddr(receiver_subaddr)
        # remove the else part, so that when receiver_id is none, we always set it as inventory account id
        # Subaddress was not specified, so route the fund to inventory account
        receiver_id = get_account(account_name=INVENTORY_ACCOUNT_NAME).id

Reply:
I suggest we move each if block into it's own method, and always returns, instead of let the logic continue until hit the last add_transaction, super hard to trace the variables.

Reply:
We want to distinguish between two cases: when receiver subaddress was None or receiver subaddress was not None but not found in ref wallet. If it was none (i.e. metadata didn't include the subaddress at all), then we want to send to inventory account. Otherwise subaddress was specified, but we couldn't find a corresponding user for it, so we send back a refund, which is handled in a separate block

Reply:
if that's the case, then we should explicitly write code for it:
```
if case 1:
       do xxxx
elif case 2:
       do yyyy
```

current implementation has no way to figure that out.

Reply:
also, even you want to refund, technically, you still need set receiver_id to inventory account first.
because the transaction need to be refunded is a fact happened onchain, it means this vasp received the money, if we don't save it to inventory account, then it means it money is not booked, then when you do refund, which is a transaction move money out of inventory account to an external account, the number won't balanced.